kep/217-resource-limit-aware-scorcing/README.md

```diff
@@ -0,0 +1,148 @@
+
+node1's resource limit is oversubscribed with a limit to allocatable ratio 14/8.  It will have a high risk of resource contention.
+
+**Proposed scheduler**
+
+If we consider node1's resource limit is already oversubscribed and hence place pod5 on node 2 instead, the result will be 
+- node1: total request/allocatable = 4/8, total limit/allocatable = 10/8  
+- node2: total request/allocatable = 6/8, total limit/allocatable = 9/8
+
+node1’s total resource limit remains the same and node2 has a total resource limit of 9. node2 will have a lower risk of resource contention with pod5 running on it. 
+
+## Terms
+
+## Proposal
+
+A custom scheduler scoring plugin scores a node by considering existing pods’ resource limits to the node's allocatable ratio on the node. It attempts to “evenly” spread or distribute pod resource limits across nodes in a cluster. 
+
+## Design Details
+
+When choosing a node to place a pod, the custom scoring plugin prefers the node that has the smallest total resource limit ratio on a node. The scoring function favors a node with the smallest over-subscription ratio.
+
+   `node limit ratio = sum of existing pods' limits on node / node’s allocatable resource`
+   
+1. Calculate each node’s raw score. A node's raw score will be negative for an over-subscription node.
+    
+    - `RawScore = (allocatable - limit) / allocatable`, where limit is the sum of existing pods on the node and the new pod's resource limits.
+    
+    - For multiple resources, the raw score is the weighted sum of all allocatable resources. The resources and their weights are configurable as the plugin arguments. Like the other `NodeResources` scoring plugins, the plugin supports standard resource types: CPU, Memory, Ephemeral-Storage plus any Scalar-Resources. 
+
+2. Normalized the raw score to `[MinNodeScore , MaxNodeScore]` (e.g., [0, 100]) after all nodes are scored.
+
+    - `NormalizedScore = MinNodeScore+(RawScore-LowestRawScore)/(HighestRawScore-LowestRawScore) *(MaxNodScore - MinNodeScore)`
+
+3. Choose the node with the highest score.
+
+## Implementation
+
+- A node's allocatable can be directly obtained from `NodeInfo.Allocatable`. 
+
+- A node's resource limit of all running pods on the node is calculated from `NodeInfo.Pods`.
+
+```
+// calculateAllocatedLimit returns the allocated resource limits on a node
+ func calculateAllocatedLimit(node *framework.NodeInfo, resource v1.ResourceName) int64 {
+ 	var limit int64
+ 	for _, pod := range node.Pods {
+ 		for _, container := range pod.Pod.Spec.Containers {
+ 			limit += getLimitForResource(resource, &container.Resources.Limits)
+ 		}
+ 	}
+ 	return limit
+ }
+
+  // GetLimitForResource returns a specified resource limit for a container
+ func getLimitForResource(resource v1.ResourceName, limits *v1.ResourceList) int64 {
+ 	switch resource {
+ 	case v1.ResourceCPU:
+ 		return limits.Cpu().MilliValue()
+ 	case v1.ResourceMemory:
+ 		return limits.Memory().Value()
+ 	default:
```

Comment:
Ensure the documented core resources like huge-pages, ephemeral-storage are also handled here. (just be consistent with the upstream logic)

Reply:
Added ephemeral-storage. Is huge-page part of `ScalarResource`? I don't find an example in the upstream plugins?

Reply:
Ah yes, it's categorized as `ScalarResource`. ([IsScalarResourceName()](https://github.com/kubernetes/kubernetes/blob/48dfa2a55420fc5078e993ac863d566962873927/pkg/scheduler/util/utils.go#L145))

Reply:
It's then covered. I added it to the description.